frontend/src/javascript/utils/codeUtils.js

```diff
@@ -59,37 +60,82 @@ export function monacoChangeToLegacyFormat(code, change) {
     return result;
 }
 
-function handleVarTrackerMove(change, varObj) {
+function isLegalVariableName(text) {
+    return /^[a-zA-Z_][a-zA-Z_0-9]*$/.test(text);
+}
+
+function isLegalVariableCharacter(char) {
+    let code = char.charCodeAt(0);
+    return (code > 47 && code < 58) || (code > 64 && code < 91) || (code > 96 && code < 123) || code == 95;
+}
+
+function isDigit(char) {
+    return char >= "0" && char <= "9";
+}
+
+function expandLeft(varObj, project) {
+    let i = varObj.start - 1;
+    while (i >= 0 && isLegalVariableCharacter(project.code[i])) i--;
+    i++;
+    while (i <= varObj.end && isDigit(project.code[i])) i++;
+    varObj.start = i;
+}
+
+function expandRight(varObj, project) {
+    let i = varObj.end;
+    while (i < project.code.length && isLegalVariableCharacter(project.code[i])) i++;
+    varObj.end = i;
+}
+
+function findFirstLegalVariableName(text) {
+    const regex = /[a-zA-Z_][a-zA-Z_0-9]*/g;
+    let match = regex.exec(text);
+    if (match == null) return null;
+    return { start: match.index, end: regex.lastIndex };
+}
+
+function handleVarTrackerMove(change, varObj, project) {
     if (varObj == null) return;
 
-    if (areIntervalsIntersect(varObj.start, varObj.end, change.start, change.end)) {
-        if (isSubinterval(varObj.start, varObj.end, change.start, change.end)) {
-            varObj.end += change.size;
-            if (varObj.end - varObj.start <= 0) return "Delete";
-            return "Rename";
-        }
-        return "Delete";
+    let originalPos = { start: varObj.start, end: varObj.end };
+
+    let newVarRange = applyChangeOnInterval(varObj.start, varObj.end, change);
+    if (newVarRange.end - newVarRange.start <= 0) return "Delete";
+    varObj.start = newVarRange.start;
+    varObj.end = newVarRange.end;
+
+    if (!areIntervalsIntersectOrTouching(originalPos.start, originalPos.end, change.start, change.end)) {
+        return;
     }
 
-    if (change.end <= varObj.start) {
-        varObj.start += change.size;
-        varObj.end += change.size;
+    let potentialName = project.code.substring(varObj.start, varObj.end);
+    if (!isLegalVariableName(potentialName)) {
```

Comment:
Ten if to jest takie zabezpieczenie przed jakimiś nieobsłużonymi przypadkami, czy np. obsłużenie tego przypadku "jeżeli zmiana jest jednoczesnym usunięciem całej zmiennej co do znaku (czyli jeśli była zaznaczona) i zmiana dodaje chociaż 1 znak (czyli próbujemy zmienić nazwę zmiennej) to śledzenie zmiennej nie usuwa się"? Ja bym to widział bardziej na zasadzie: "Jeżeli jest wiecej niż jeden kandydat na zmienną w tym przedziale to Delete(prawdopodobnie użytkownik wkleił jakiś większy fragment kodu), a jeżeli jest tylko jeden kandydat to rename (z ewentualną poprawką start i end)"